build/Dockerfile

```diff
@@ -127,50 +132,95 @@ RUN if [[ -z "${WHEEL_VERSION}" ]]; \
     fi && \
     ls /tmp/*.whl >/tmp/bdist_name
 
+## Stages for dev images ######################################
+FROM files-common as dev
+ARG USER
+ARG USER_UID
+ARG ENABLE_AIM
+
+# Install from the wheel / optionals deps, pytest, etc
+RUN --mount=type=cache,target=/home/${USER}/.cache/pip,uid=${USER_UID} \
+    python -m pip install --user wheel tox pytest && \
+    python -m pip install --user "$(head /tmp/bdist_name)" && \
+    python -m pip install --user "$(head /tmp/bdist_name)[flash-attn]" && \
+    python -m pip install --user "$(head /tmp/bdist_name)[dev]" && \
+    if [[ "${ENABLE_AIM}" == "true" ]]; then \
+        python -m pip install --user "$(head /tmpbdist_name)[aim]"; \
+    fi && \
+    python -m pip uninstall wheel build -y && \
+    rm $(head bdist_name) /tmp/bdist_name
+
+RUN if [[ "${ENABLE_AIM}" == "true" ]] ; then \
+        touch /.aim_profile && \
+        chmod -R 777 /.aim_profile; \
+    fi
+
+# Create the directory for vscode-server; this directory has to be pre-created
+# such that the user can write to it, otherwise we can't attach a vscode instance
+# to it.
+RUN mkdir -p /app/.vscode-server && \
+    chown $USER:0 /app/.vscode-server
+
+WORKDIR /app
+USER ${USER}
+
+# Unit tests, build infrastructure, common scripts
+COPY --from=files-common /app/ /app/
+COPY --from=files-common /.cache/ /.cache/
+COPY tests /app/tests
+COPY tox.ini /app/
+COPY Makefile /app/
+COPY scripts /app/scripts
+
+ENV FSDP_DEFAULTS_FILE_PATH="/app/accelerate_fsdp_defaults.yaml"
+ENV SET_NUM_PROCESSES_TO_NUM_GPUS="True"
+ENV PYTHONPATH="/home/${USER}/.local/lib/python${PYTHON_VERSION}/site-packages:/app"
+
+## Stages for release images ##################################
+FROM files-common as python-rel-installations
+ARG USER
+ARG USER_UID
+ARG ENABLE_AIM
+
 # Install from the wheel
+# TODO - probably a good idea to install most stuff in common and copy it out
+# in both this stage and dev.
```

Comment:
Or similarly, could the dev stage come after the python building stage? But the dev stage adds the additional tools and then release stage removes the unnecessary parts? I guess it may be hard to install the dev tools and selectively remove them...